src/core/net/dnssd.hpp

```diff
@@ -0,0 +1,281 @@
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for infrastructure DNS-SD (mDNS) platform.
+ */
+
+#ifndef DNSSD_HPP_
+#define DNSSD_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_PLATFORM_DNSSD_ENABLE
+
+#include <openthread/platform/dnssd.h>
+
+#include "common/clearable.hpp"
+#include "common/locator.hpp"
+#include "common/non_copyable.hpp"
+#include "net/ip6_address.hpp"
+
+namespace ot {
+
+/**
+ * @addtogroup core-dns
+ *
+ * @brief
+ *   This module includes definitions for DNS-SD (mDNS) platform.
+ *
+ * @{
+ *
+ */
+
+extern "C" void otPlatDnssdStateHandleStateChange(otInstance *aInstance);
+
+/**
+ * Represents DNS-SD (mDNS) platform.
+ *
+ */
+class Dnssd : public InstanceLocator, private NonCopyable
+{
+    friend void otPlatDnssdStateHandleStateChange(otInstance *aInstance);
+
+public:
+    /**
+     * Represents state of DNS-SD platform.
+     *
+     */
+    enum State : uint8_t
+    {
+        kStopped = OT_PLAT_DNSSD_STOPPED, ///< Stopped and unable to register any service or host.
+        kReady   = OT_PLAT_DNSSD_READY    ///< Running and ready to register service or host.
+    };
+
+    typedef otPlatDnssdRequestId        RequestId;        ///< A request ID.
+    typedef otPlatDnssdRegisterCallback RegisterCallback; ///< The registration request callback
+
+    class Host : public otPlatDnssdHost, public Clearable<Host> ///< Host information.
+    {
+    };
+
+    class Service : public otPlatDnssdService, public Clearable<Service> ///< Service information.
+    {
+    };
+
+    class Key : public otPlatDnssdKey, public Clearable<Key> ///< Key information
+    {
+    };
+
+    /**
+     * Represents a range of `RequestId` values.
+     *
+     * The range is stored using start and end ID values. The implementation handles the case when ID values roll over.
+     *
+     */
+    struct RequestIdRange : public Clearable<RequestIdRange>
```

Comment:
Exactly.

It represents a range of "Request ID" (from a start ID value up to to an end value).

- It is relatively small data structure so we can include it in every SRP `Host` entry (without adding too much size overhead). 
- We use it when we get a callback for a request ID to skip over some `Host` (and `Service` entries). If none of services of a `Host` is waiting for the ID, we can skip iterating through its service.
- And since the request IDs are allocated sequentially, the request IDs associated with same `Host` often fit in a short range.